src/miner.cpp

```diff
@@ -671,99 +613,61 @@ void BitcoinMiner(CWallet* pwallet, bool fProofOfStake)
+                MilliSleep(2000);
+                continue;
             }
+
+        } else if ((pindexPrev->nHeight - 6) > Params().LAST_POW_BLOCK()) {
+            // Late PoW: run for a little while longer, just in case there is a rewind on the chain.
+            LogPrintf("%s: Exiting PoW Mining Thread at height: %d\n", __func__, pindexPrev->nHeight);
+            return;
        }
 
         //
         // Create new block
         //
         unsigned int nTransactionsUpdatedLast = mempool.GetTransactionsUpdated();
 
-        std::unique_ptr<CBlockTemplate> pblocktemplate(
-                fProofOfStake ? CreateNewBlock(CScript(), pwallet, fProofOfStake) : CreateNewBlockWithKey(reservekey, pwallet)
-                        );
-        if (!pblocktemplate.get())
-            continue;
-
+        std::unique_ptr<CBlockTemplate> pblocktemplate((fProofOfStake ?
+                                                        CreateNewBlock(CScript(), pwallet, fProofOfStake) :
+                                                        CreateNewBlockWithKey(reservekey, pwallet)));
+        if (!pblocktemplate.get()) continue;
         CBlock* pblock = &pblocktemplate->block;
-        IncrementExtraNonce(pblock, pindexPrev, nExtraNonce);
 
-        //Stake miner main
+        // POS - block found: process it
         if (fProofOfStake) {
-            LogPrintf("CPUMiner : proof-of-stake block found %s \n", pblock->GetHash().ToString().c_str());
-            if (pblock->IsZerocoinStake()) {
-                //Find the key associated with the zerocoin that is being staked
-                libzerocoin::CoinSpend spend = TxInToZerocoinSpend(pblock->vtx[1].vin[0]);
-                CBigNum bnSerial = spend.getCoinSerialNumber();
-                CKey key;
-                if (!pwallet->GetZerocoinKey(bnSerial, key)) {
-                    LogPrintf("%s: failed to find zPIV with serial %s, unable to sign block\n", __func__, bnSerial.GetHex());
-                    continue;
-                }
-
-                //Sign block with the zPIV key
-                if (!SignBlockWithKey(*pblock, key)) {
-                    LogPrintf("%s: Signing new block with zPIV key failed \n", __func__);
-                    continue;
-                }
-            } else if (!SignBlock(*pblock, *pwallet)) {
-                LogPrintf("%s: Signing new block with UTXO key failed \n", __func__);
-                continue;
-            }
-
             LogPrintf("%s : proof-of-stake block was signed %s \n", __func__, pblock->GetHash().ToString().c_str());
             SetThreadPriority(THREAD_PRIORITY_NORMAL);
             if (!ProcessBlockFound(pblock, *pwallet, reservekey)) {
                 LogPrintf("%s: New block orphaned\n", __func__);
                 continue;
             }
             SetThreadPriority(THREAD_PRIORITY_LOWEST);
-
             continue;
         }
 
+        // POW - miner main
+        IncrementExtraNonce(pblock, pindexPrev, nExtraNonce);
```

Comment:
Can we put this under the else path of the  `if (fProofOfStake)` block that it's above.

First glance is hard to get it, the whole `if (fProofOfStake)` block that it's above basically start the flow all over independently on what happen.. lot of ugly `continue` there.